File: Euler14.cpp
#include <stdio.h>
#include <map>

std::map<long long int, int> mymap;


int collatz(long long int index)
{
  // printf ("Collatz (%lld)\n", index);
  int & result = mymap[index];

  if (result != 0)
  {
    return result;
  }
  else
  {
    result = index % 2 == 0 ? collatz(index / 2) : collatz(index * 3 + 1);
    ++result;
    return result;
  }
}

int main (void)
{
  mymap[1] = 1;
  int max = 0;
  int max_index = 0;
  for (int i = 1; i <= 1000000; ++i)
  {
    // printf ("%d\n", i);
    int curr = collatz(i);
    if (curr > max) { max = curr; max_index = i; }
  }

  printf ("%d, %d\n", max, max_index);
}

File: Euler14.py
from array import array

cache = dict()

cache[1] = 1;

def collatz(i: int):
    if i in cache:
        return cache[i]
    else:
        result = collatz(int(i/2)) if i % 2 == 0 else collatz(i * 3 + 1)
        result += 1
        cache[i] = result
        return result


max = (0, 0)

for i in range (1, 1000000):
    c = collatz(i)
    if c > max[0]:
        max = (c, i)

print (str(max))
